### PR TITLE
[CPP] Add missing metadata on Arrow schemas returned by Flight SQL #11999 

### PR DESCRIPTION
**Pull request by jcralmeida**
This add an auxiliary class FlightSqlColumnMetadata meant to read and write known metadata for Arrow schema fields, such as CATALOG_NAME, SCHEMA_NAME, TABLE_NAME, PRECISION, SCALE, IS_AUTO_INCREMENT, IS_CASE_SENSITIVE, IS_READ_ONLY and IS_SEARCHABLE.

### CODE REVIEW

File: cpp/src/arrow/flight/sql/column_metadata.h [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+ private:
+  std::shared_ptr<arrow::KeyValueMetadata> metadata_map_;
+  explicit ColumnMetadata(std::shared_ptr<arrow::KeyValueMetadata> metadata_map);
+
+ public:
+  class ColumnMetadataBuilder;
+
+  /// \brief Constant variable to hold the value of the key that
+  ///        will be used in the KeyValueMetadata class.
+  static const char* CATALOG_NAME;
```
Comment (lidavidm) [inline cpp/src/arrow/flight/sql/column_metadata.h:39]: nit: constants should be named `kCatalogName` according to the [style guide](https://google.github.io/styleguide/cppguide.html#Constant_Names)

File: cpp/src/arrow/flight/sql/column_metadata.cc [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/flight/sql/column_metadata.h"
+
+#include <utility>
+
+namespace {
```
Comment (lidavidm) [inline cpp/src/arrow/flight/sql/column_metadata.cc:22]: nit: the anonymous namespace could probably be nested into the arrow::flight::sql namespace below

File: cpp/src/arrow/flight/sql/column_metadata.h [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  /// \brief Constant variable to hold the value of the key that
+  ///        will be used in the KeyValueMetadata class.
+  static const char* IS_SEARCHABLE;
+
+  /// \brief Static initializer.
+  static ColumnMetadataBuilder Builder();
+
+  /// \brief  Return the catalog name set in the KeyValueMetadata.
+  /// \return The catalog name.
+  arrow::Result<std::string> GetCatalogName();
```
Comment (lidavidm) [inline cpp/src/arrow/flight/sql/column_metadata.h:70]: Can these be declared `const`?

File: cpp/src/arrow/flight/sql/column_metadata.cc (hunk 2 above)
Comment (jcralmeida) [reply cpp/src/arrow/flight/sql/column_metadata.cc:25]: Fixed

File: cpp/src/arrow/flight/sql/column_metadata.h (hunk 3 above)
Comment (jcralmeida) [reply cpp/src/arrow/flight/sql/column_metadata.h]: Fixed

**Comment by jcralmeida** [comment]
Closed this pull request because both implementation JAVA and C++ are now in the same Pull Request

https://github.com/apache/arrow/pull/11999

**Pull request closed without merging**